### ISSUE TITLE
Login failed and returned `401`

### ISSUE THREAD

**Issue by kenkangxgwe** [millejoh/emacs-ipython-notebook#491]
## Problem description

The login failed and returned error code `401`.
Seems that the token is not parsed correctly?

```
ein: [error] Login to http://127.0.0.1:8888 failed, error-thrown (error http 401), raw-header HTTP/1.1 401 Unauthorized
Server: TornadoServer/4.5.3
Content-Type: text/html; charset=UTF-8
Date: Sun, 17 Mar 2019 05:21:44 GMT
Content-Security-Policy: frame-ancestors 'self'; report-uri /api/security/csp-report
Content-Length: 8942
Set-Cookie: username-127-0-0-1-8888=""; expires=Sat, 17 Mar 2018 05:21:44 GMT; Path=/
```

## Steps to reproduce the problem

<kbd>M-x</kbd>`ein:run`

## System info:

```cl
("EIN system info"
 :emacs-version "GNU Emacs 26.1 (build 1, x86_64-w64-mingw32)
 of 2018-05-28"
 :emacs-bzr-version "ae330c90e2cb41554c42a9d0154a00e084d5bb04"
 :window-system w32
 :emacs-variant nil
 :os
 (:uname "MINGW32_NT-6.2 MYUSERNAME 1.0.13(0.47/3/2) 2010-01-27 23:29 i686 Msys
"
         :lsb-release nil)
 :notebook
 "Name: notebook
Version: 5.4.0
Summary: A web-based notebook environment for interactive computing
Home-page: http://jupyter.org
Author: Jupyter Development Team
Author-email: jupyter@googlegroups.com
License: BSD
Location: d:\\programs\\anaconda3\\lib\\site-packages
Requires: jinja2, tornado, ipython-genutils, traitlets, jupyter-core, jupyter-client, nbformat, nbconvert, ipykernel, Send2Trash, terminado
Required-by: widgetsnbextension, jupyterlab, jupyterlab-launcher, jupyter
"
 :ipython "6.2.1
"
 :image-types
 (svg png gif tiff jpeg xpm xbm pbm)
 :image-types-available
 (svg png gif tiff jpeg xpm xbm pbm)
 :request
 (:backend curl)
 :ein
 (:version "0.15.0-20190316.414"
           :source-dir "c:/Users/kenkangxgwe/.emacs.d/elpa/develop/ein-20190316.414/")
 :lib
 ((:name "websocket"
         :path "~/.emacs.d/elpa/develop/websocket-20180423.16/websocket.elc"
         :featurep t
         :version-var websocket-version
         :version "1.9")
  (:name "request"
         :path "~/.emacs.d/elpa/develop/request-20181129.1138/request.elc"
         :featurep t
         :version-var request-version
         :version "0.3.0")
  (:name "auto-complete"
         :path "~/.emacs.d/elpa/develop/auto-complete-20170125.245/auto-complete.elc"
         :featurep t
         :version-var nil
         :version nil)
  (:name "popup"
         :path "~/.emacs.d/elpa/develop/popup-20160709.1429/popup.elc"
         :featurep t
         :version-var popup-version
         :version "0.5.3")
  (:name "fuzzy"
         :path "~/.emacs.d/elpa/develop/fuzzy-20150730.337/fuzzy.elc"
         :featurep nil
         :version-var nil
         :version nil)
  (:name "pos-tip"
         :path "~/.emacs.d/elpa/develop/pos-tip-20150318.1513/pos-tip.elc"
         :featurep t
         :version-var pos-tip-version
         :version "0.4.6")
  (:name "python"
         :path "d:/Programs/msys64/mingw64/share/emacs/26.1/lisp/progmodes/python.elc"
         :featurep t
         :version-var nil
         :version nil)
  (:name "markdown-mode"
         :path "~/.emacs.d/elpa/develop/markdown-mode-20190305.319/markdown-mode.elc"
         :featurep nil
         :version-var nil
         :version nil)
  (:name "helm"
         :path "~/.emacs.d/elpa/develop/helm-core-20190205.1337/helm.elc"
         :featurep t
         :version-var nil
         :version nil)))
```

## `*ein:log-all*`

```
01:21:06:997: [info] ein:jupyter-server--run: d:/Programs/Anaconda3/Scripts/jupyter.exe notebook --notebook-dir=d:/Projects/ --no-browser @#<buffer *spacemacs*>
01:21:14:278: [error] Login to http://127.0.0.1:8888 failed, error-thrown (error http 401), raw-header HTTP/1.1 401 Unauthorized
Server: TornadoServer/4.5.3
Content-Type: text/html; charset=UTF-8
Date: Sun, 17 Mar 2019 05:21:14 GMT
Content-Security-Policy: frame-ancestors 'self'; report-uri /api/security/csp-report
Content-Length: 8942
Set-Cookie: username-127-0-0-1-8888=""; expires=Sat, 17 Mar 2018 05:21:14 GMT; Path=/
 @#<buffer *spacemacs*>
01:21:28:016: [info] Stopped Jupyter notebook server. @#<buffer *spacemacs*>
01:21:36:605: [info] ein:jupyter-server--run: d:/Programs/Anaconda3/Scripts/jupyter.exe notebook --notebook-dir=d:/Projects/ --no-browser @#<buffer *spacemacs*>
01:21:43:979: [debug] Login attempt #0 in response to nil from http://127.0.0.1:8888. @#<buffer *spacemacs*>
01:21:44:289: [error] Login to http://127.0.0.1:8888 failed, error-thrown (error http 401), raw-header HTTP/1.1 401 Unauthorized
Server: TornadoServer/4.5.3
Content-Type: text/html; charset=UTF-8
Date: Sun, 17 Mar 2019 05:21:44 GMT
Content-Security-Policy: frame-ancestors 'self'; report-uri /api/security/csp-report
Content-Length: 8942
Set-Cookie: username-127-0-0-1-8888=""; expires=Sat, 17 Mar 2018 05:21:44 GMT; Path=/
 @#<buffer *spacemacs*>
01:21:44:331: [debug] ein:notebooklist-login--complete STATUS: 401 DATA: (:bad-page 5348) @#<buffer *spacemacs*>
```
## `*request-log*`

```
[debug] REQUEST
[debug] Run: curl --silent --include --location --compressed --cookie c:/Users/kenkangxgwe/.emacs.d/.cache/request/curl-cookie-jar --cookie-jar c:/Users/kenkangxgwe/.emacs.d/.cache/request/curl-cookie-jar --write-out \n(:num-redirects %{num_redirects} :url-effective %{url_effective}) --data-binary @c:/Users/KENKAN~1/AppData/Local/Temp/emacs-requestO2fFlH --header X-XSRFTOKEN: 2|3e1bce27|abfc2558ae31bb27b741c4b5c2ff3328|1552610680 http://127.0.0.1:8888/login
[debug] Start timer: timeout=6.3 sec
[debug] REQUEST--CURL-CALLBACK event = finished

[debug] REQUEST--CURL-CALLBACK proc = #<process request curl>
[debug] REQUEST--CURL-CALLBACK buffer = #<buffer  *request curl*>
[debug] REQUEST--CURL-CALLBACK symbol-status = nil
[debug] REQUEST--CALLBACK
[debug] (buffer-string) =
HTTP/1.1 401 Unauthorized

Server: TornadoServer/4.5.3

Content-Type: text/html; charset=UTF-8

Date: Sun, 17 Mar 2019 05:21:44 GMT

Content-Security-Policy: frame-ancestors 'self'; report-uri /api/security/csp-report

Content-Length: 8942

Set-Cookie: username-127-0-0-1-8888=""; expires=Sat, 17 Mar 2018 05:21:44 GMT; Path=/



<!DOCTYPE HTML>
<html>

<head>
    <meta charset="utf-8">

    <title>Jupyter Notebook</title>
    <link id="favicon" rel="shortcut icon" type="image/x-icon" href="/static/base/images/favicon.ico?v=97c6417ed01bdc0ae3ef32ae4894fd03">
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <link rel="stylesheet" href="/static/components/jquery-ui/themes/smoothness/jquery-ui.min.css?v=9b2c8d3489227115310662a343fce11c" type="text/css" />
    <link rel="stylesheet" href="/static/components/jquery-typeahead/dist/jquery.typeahead.min.css?v=7afb461de36accb1aa133a1710f5bc56" type="text/css" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    

    <link rel="stylesheet" href="/static/style/style.min.css?v=47782e517c98a53adb514cbefb4528f2" type="text/css"/>
    
<link rel="stylesheet" href="/static/auth/css/override.css?v=19ec59d2c4f1203c49fe47462028cd9a" type="text/css" />

    <link rel="stylesheet" href="/custom/custom.css" type="text/css" />
    <script src="/static/components/es6-promise/promise.min.js?v=f004a16cb856e0ff11781d01ec5ca8fe" type="text/javascript" charset="utf-8"></script>
    <script src="/static/components/preact/index.js?v=00a2fac73c670ce39ac53d26640eb542" type="text/javascript"></script>
    <script src="/static/components/proptypes/index.js?v=c40890eb04df9811fcc4d47e53a29604" type="text/javascript"></script>
    <script src="/static/components/preact-compat/index.js?v=f865e990e65ad27e3a2601d8adb48db1" type="text/javascript"></script>
    <script src="/static/components/requirejs/require.js?v=6da8be361b9ee26c5e721e76c6d4afce" type="text/javascript" charset="utf-8"></script>
    <script>
      require.config({
          
          urlArgs: "v=20190317012138",
          
          baseUrl: '/static/',
          paths: {
            'auth/js/main': 'auth/js/main.min',
            custom : '/custom',
            nbextensions : '/nbextensions',
            kernelspecs : '/kernelspecs',
            underscore : 'components/underscore/underscore-min',
            backbone : 'components/backbone/backbone-min',
            jed: 'components/jed/jed',
            jquery: 'components/jquery/jquery.min',
            json: 'components/requirejs-plugins/src/json',
            text: 'components/requirejs-text/text',
            bootstrap: 'components/bootstrap/js/bootstrap.min',
            bootstraptour: 'components/bootstrap-tour/build/js/bootstrap-tour.min',
            'jquery-ui': 'components/jquery-ui/ui/minified/jquery-ui.min',
            moment: 'components/moment/min/moment-with-locales',
            codemirror: 'components/codemirror',
            termjs: 'components/xterm.js/dist/xterm',
            typeahead: 'components/jquery-typeahead/dist/jquery.typeahead.min',
          },
          map: { // for backward compatibility
              "*": {
                  "jqueryui": "jquery-ui",
              }
          },
          shim: {
            typeahead: {
              deps: ["jquery"],
              exports: "typeahead"
            },
            underscore: {
              exports: '_'
            },
            backbone: {
              deps: ["underscore", "jquery"],
              exports: "Backbone"
            },
            bootstrap: {
              deps: ["jquery"],
              exports: "bootstrap"
            },
            bootstraptour: {
              deps: ["bootstrap"],
              exports: "Tour"
            },
            "jquery-ui": {
              deps: ["jquery"],
              exports: "$"
            }
          },
          waitSeconds: 30,
      });

      require.config({
          map: {
              '*':{
                'contents': 'services/contents',
              }
          }
      });

      // error-catching custom.js shim.
      define("custom", function (require, exports, module) {
          try {
              var custom = require('custom/custom');
              console.debug('loaded custom.js');
              return custom;
          } catch (e) {
              console.error("error loading custom.js", e);
              return {};
          }
      })

    document.nbjs_translations = {"domain": "nbjs", "locale_data": {"nbjs": {"": {"domain": "nbjs"}}}};
    </script>

    
    

</head>

<body class=""
 
  
 
dir="ltr">

<noscript>
    <div id='noscript'>
      Jupyter Notebook requires JavaScript.<br>
      Please enable it to proceed. 
  </div>
</noscript>

<div id="header">
  <div id="header-container" class="container">
  <div id="ipython_notebook" class="nav navbar-brand"><a href="/tree" title='dashboard'>
      <img src='/static/base/images/logo.png?v=641991992878ee24c6f3826e81054a0f' alt='Jupyter Notebook'/>
  </a></div>

  
  
  
  
  
  


  
  
  </div>
  <div class="header-bar"></div>

  
  
</div>

<div id="site">


<div id="ipython-main-app" class="container">

    
    
    <div class="row">
    <div class="navbar col-sm-8">
      <div class="navbar-inner">
        <div class="container">
          <div class="center-nav">
          
            <p class="navbar-text nav">Password or token:</p>
          
            <form action="/login?next=%2F" method="post" class="navbar-form pull-left">
              <input type="hidden" name="_xsrf" value="2|f9a50e0f|6c42e570698f7b0f70ff049d0541f300|1552610680"/>
              <input type="password" name="password" id="password_input" class="form-control">
              <button type="submit" id="login_submit">Log in</button>
            </form>
          </div>
        </div>
      </div>
    </div>
    </div>
    
    
      <div class="row">
      
        <div class="message error">
           Invalid credentials
        </div>
      
      </div>
    
    
    
    <div class="col-sm-6 col-sm-offset-3 text-left rendered_html">
      <h3>
        Token authentication is enabled
      </h3>
      <p>
        If no password has been configured, you need to open the notebook
        server with its login token in the URL, or paste it above.
        This requirement will be lifted if you
        <b><a href='https://jupyter-notebook.readthedocs.io/en/stable/public_server.html'>
            enable a password</a></b>.
      </p>
      <p>
        The command:
        <pre>jupyter notebook list</pre>
        will show you the URLs of running servers with their tokens,
        which you can copy and paste into your browser. For example:
      </p>
      <pre>Currently running servers:
http://localhost:8888/?token=c8de56fa... :: /Users/you/notebooks
</pre>
      <p>
        or you can paste just the token value into the password field on this
        page.
      </p>
      <p>
        See
        <b><a
         href='https://jupyter-notebook.readthedocs.io/en/stable/public_server.html'>
                the documentation on how to enable a password</a>
        </b>
        in place of token authentication,
        if you would like to avoid dealing with random tokens.
      </p>
      <p>
        Cookies are required for authenticated access to notebooks.
      </p>
      
        <h3>Setup a Password</h3>
        <p> You can also setup a password by entering your token and a new password
        on the fields below:</p>
        <form action="/login?next=%2F" method="post" class="">
                <input type="hidden" name="_xsrf" value="2|f9a50e0f|6c42e570698f7b0f70ff049d0541f300|1552610680"/>
          <div class="form-group">
            <input type="password" name="password" id="password_input" class="form-control" placeholder="Token">
          </div>
          <div class="form-group">
            <input type="password" name="new_password" id="new_password_input"
            class="form-control" placeholder="New password" required>
          </div>
          <div class="form-group">
            <button type="submit" id="login_submit">Log in and set new password</button>
          </div>
        </form>
      

    </div>
    
    
</div>


</div>








<script type="text/javascript">
  require(["auth/js/main"], function (auth) {
    auth.login_main();
  });
</script>



<script type='text/javascript'>
  function _remove_token_from_url() {
    if (window.location.search.length <= 1) {
      return;
    }
    var search_parameters = window.location.search.slice(1).split('&');
    for (var i = 0; i < search_parameters.length; i++) {
      if (search_parameters[i].split('=')[0] === 'token') {
        // remote token from search parameters
        search_parameters.splice(i, 1);
        var new_search = '';
        if (search_parameters.length) {
          new_search = '?' + search_parameters.join('&');
        }
        var new_url = window.location.origin + 
                      window.location.pathname + 
                      new_search + 
                      window.location.hash;
        window.history.replaceState({}, "", new_url);
        return;
      }
    }
  }
  _remove_token_from_url();
</script>
</body>

</html>
[debug] REQUEST-RESPONSE--CANCEL-TIMER
[debug] -CLEAN-HEADER
[debug] -CUT-HEADER
[debug] error-thrown = (error http 401)
[debug] -PARSE-DATA
[debug] parser = ein:notebooklist-login--parser
[debug] data = (:bad-page 5348)
[debug] symbol-status = error
[debug] Executing error callback.
[debug] Executing complete callback.
```

**Comment by dickmao** [comment]
This is a problem.  Can you:

1. `M-x ein:dev-start-debug`
2. `M-x ein:run`.  Should reproduce error.
3. Switch to `*scratch*` buffer and copy the following:
```
(ein:notebooklist-token-or-password (ein:url 8888))

(ein:query-singleton-ajax 'blah (ein:url 8888 "login") :data (concat "password=" (url-hexify-string (ein:notebooklist-token-or-password (ein:url 8888)))) :parser #'ein:notebooklist-login--parser)

(ein:query-singleton-ajax 'blah (ein:url 8888 "login") :data (concat "password=" (ein:notebooklist-token-or-password (ein:url 8888))) :parser #'ein:notebooklist-login--parser)
```
4. Go to the end of the first expression and `C-x C-e`.  Ensure it does not return nil.
5. Go to end of the second expression and `C-x C-e`.  Now `M-x ein:log-pop-to-request-buffer` and go to end of buffer.  Ensure the error is reproduced.
6. Go back to `*scratch*` and go to end of third expression `C-x C-e`.  Now `M-x ein:log-pop-to-request-buffer` and go to end of buffer.  Success or still error?

**Comment by kenkangxgwe** [comment]
Here are the results
```elisp
(ein:notebooklist-token-or-password (ein:url 8888))

> "c2883781ea81734e37312cedb1aebefc35ce46acbe33a235"


(equal (ein:notebooklist-token-or-password (ein:url 8888)) (url-hexify-string (ein:notebooklist-token-or-password (ein:url 8888))))

> t


(ein:query-singleton-ajax 'blah (ein:url 8888 "login") :data (concat "password=" (url-hexify-string (ein:notebooklist-token-or-password (ein:url 8888)))) :parser #'ein:notebooklist-login--parser)

> #s(request-response nil nil nil nil nil "http://127.0.0.1:8888/login" nil (:data "password=c2883781ea81734e37312cedb1aebefc35ce46acbe33a235" :parser ein:notebooklist-login--parser :headers (("X-XSRFTOKEN" . "2|3e1bce27|abfc2558ae31bb27b741c4b5c2ff3328|1552610680")) :error #[128 "\302\300\303\301\"\"\207" [request-default-error-callback ("http://127.0.0.1:8888/login") apply append] 6 "

(fn &rest ARGS2)"] :url "http://127.0.0.1:8888/login" :response #0) #<buffer  *request curl*> nil nil curl ("c:/Users/KENKAN~1/AppData/Local/Temp/emacs-requestg4yDPK"))
```

And the `*request-log*` buffer was still error

```
[debug] REQUEST-RESPONSE--CANCEL-TIMER
[debug] -CLEAN-HEADER
[debug] -CUT-HEADER
[debug] error-thrown = (error http 401)
[debug] -PARSE-DATA
[debug] parser = ein:notebooklist-login--parser
[debug] data = (:bad-page 5348)
[debug] symbol-status = error
[debug] Executing error callback.
[error] Error (error) while connecting to http://127.0.0.1:8888/login.
```

**Comment by dickmao** [comment]
I'm afraid I cannot help you.

You can try adding `--debug` to the "Ein:Jupyter Server Args" in `M-x ein:customize-group RET ein`, restart, reproduce the error, and switch to buffer `*ein:jupyter-server*` to see if jupyter will explain why it refuses your token.

Also, check you haven't any weird stuff in `$HOME/.jupyter/jupyter_notebook_config.py`.

Otherwise, I thank you for your interest in EIN, and farewell.

**Comment by HongpengM** [comment]
I've met the same problem using mac-emacs build. 
As wiki tips suggest, maybe it's your emacs's `gnu-tls` is ill-configured or not installed. 
[https://github.com/millejoh/emacs-ipython-notebook/wiki/Tips](url)

I have 2 emacs build, one with `gnu-tls` and one without.
It turns out the emacs met exactly the same problem doesn't have available `gnu-tls`.
you could check using 
`(gnutls-available-p)`

I see you are using windows build of emacs, maybe you could check `gnu-tls` part.
If it's missed, I suggest that you try adding `gnu-tls` first.

**Comment by dickmao** [comment]
That's a very good suggestion.  `gnu-tls` has quietly broken things for me in the past as well.

In the meantime, I've added a full trace dump of `curl` to the `M-x ein:log-pop-to-request-buffer` in #493

**Comment by kenkangxgwe** [comment]
After leaving it for several days,  I just upgraded the package, and the problem has gone. 

Time is the best weapon. : - )

**Comment by mlindsk** [comment]
Hi,

I've struggled tremendously with this on windows (all good on linux). Apparently it works out of the box for me on each fresh restart of the computer. If, however I close a session and want to work on another session I get the above error. If I write `jupyter notebook list` in a shell it shows up as running although I thought I disconnected. This will produce the error when trying to start a new notebook. So my workflow is just to `jupyter notebook stop` before I open a new notebook. Works like a charm.
